### PR TITLE
Bug 1275425 - Hacky workaround for issues storing astral characters.

### PR DESCRIPTION
**Pull request by jgraham**
Test names, messages, etc. may contain UTF8 characters from beyond the
Basic Multilingual Plane ("astral" characters). Unfortunately MySQL's
"utf8" character set is nothing of the sort and will only store a
maximum of three bytes per character, thus restricting it to BMP
characters. The correct fix to this is to switch to the utf8mb4
character set. Since such a change is somewhat involved, however, we
address the immediate problem with a hack.

When storing failure lines, if the operation fails for character set
related reasons, try again with any non-BMP characters replaced by a
marker of the form <U+codepoint> e.g. <U+10FFFF>.

Note further that whether or not MySQL fails here or silently replaces
each byte of the original character with a U+FFFD replacement character
depends on the value of the sql_mode setting. If this is set to
STRICT_ALL_TABLES, we get an error, otherwise silent data
loss. Therefore it is important this setting is consistent across all
environments.

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/treeherder/1512)

### CODE REVIEW

File: puppet/files/mysql/my.cnf [hunk 1]
```diff
@@ -106,7 +106,10 @@ max_binlog_size         = 100M
 # ssl-ca=/etc/mysql/cacert.pem
 # ssl-cert=/etc/mysql/server-cert.pem
 # ssl-key=/etc/mysql/server-key.pem
-
+#
+# * SQL mode
+#
+sql_mode="STRICT_ALL_TABLES"
```
Comment (edmorley) [inline puppet/files/mysql/my.cnf]: I wonder if this setting is consistent across {stage, prod, heroku-*}?

Comment (jgraham) [inline puppet/files/mysql/my.cnf]: It isn't on heroku at least. But it should be! We need some way to make sure we have the same MySQL configuration everywhere.

File: treeherder/log_parser/management/commands/store_failure_lines.py [hunk 2]
```diff
@@ -35,30 +36,58 @@ def handle(self, *args, **options):
...
+def replace_astral(log_iter):
+    for item in log_iter:
+        for key in ["test", "subtest", "message", "stack", "stackwalk_stdout",
+                    "stackwalk_stderr"]:
+            if key in item:
+                item[key] = astral_filter(item[key])
+        yield item
+
+
+filter_re = re.compile(ur"([\U00010000-\U0010FFFF])", re.U)
```
Comment (edmorley) [inline treeherder/log_parser/management/commands/store_failure_lines.py]: What does this range represent? Comments please :-)

File: treeherder/log_parser/management/commands/store_failure_lines.py [hunk 3]
```diff
@@ -35,30 +36,58 @@ def handle(self, *args, **options):
...
         with transaction.atomic():
-            FailureLine.objects.bulk_create(
-                [FailureLine(repository=repository, job_guid=job_guid, **failure_line)
-                 for failure_line in log_iter]
-            )
+            try:
+                create(repository, job_guid, log_iter)
+            except OperationalError as e:
+                logger.warning("Got OperationalError inserting failure_line")
+                retry = e.args[0] == 1366
```
Comment (edmorley) [inline treeherder/log_parser/management/commands/store_failure_lines.py]: What is this magic number? Comments please :-)

File: treeherder/log_parser/management/commands/store_failure_lines.py [hunk 4]
```diff
@@ -35,30 +36,58 @@ def handle(self, *args, **options):
 
         log_text = fetch_text(job_log_url)
 
-        if not log_text:
-            return
-
-        log_content = StringIO(log_text)
```
Comment (edmorley) [inline treeherder/log_parser/management/commands/store_failure_lines.py]: Why the switch from StringIO? Add a comment to the commit message please :-)

File: .travis.yml [hunk 5]
```diff
@@ -107,7 +109,9 @@ matrix:
       before_install:
         # Manually install mysql 5.6 since the default is v5.5.
         - sudo apt-get -qq update
-        - sudo apt-get -qq install mysql-server-5.6 mysql-client-5.6 mysql-client-core-5.6
+        - sudo apt-get -y install mysql-server-5.6 mysql-client-5.6 mysql-client-core-5.6
```
Comment (edmorley) [inline .travis.yml]: The apt-get line can be left as is :-)

**Pull request merged**